diff -Naur unpatched/run_gbs_pipeline.pl quiet.patched/run_gbs_pipeline.pl
--- unpatched/run_gbs_pipeline.pl	1970-01-01 12:00:01.000000000 +1200
+++ quiet.patched/run_gbs_pipeline.pl	2024-11-21 09:51:42.906917184 +1300
@@ -14,8 +14,8 @@
 }
 push(@fl, "$top/dist/sTASSELGBS.jar");
 my $CP = join(":", @fl);
-print $CP;
-print "\n";
+# print $CP;
+# print "\n";
 my @args = @ARGV;
 
 system "java -classpath '$CP' -Xms512m -Xmx1536m net.maizegenetics.pipeline.TasselPipeline @args";
diff -Naur unpatched/run_pipeline.pl quiet.patched/run_pipeline.pl
--- unpatched/run_pipeline.pl	1970-01-01 12:00:01.000000000 +1200
+++ quiet.patched/run_pipeline.pl	2024-11-21 09:52:18.256032952 +1300
@@ -18,7 +18,7 @@
 }
 push(@fl, "$top/dist/sTASSEL.jar");
 my $CP = join(":", @fl);
-print $CP . "\n";
+# print $CP . "\n";
 
 # Scan @ARGV for Java memory arguments, and put rest in @args
 my $java_mem_min_default = "-Xms512m";
diff -Naur unpatched/start_tassel.pl quiet.patched/start_tassel.pl
--- unpatched/start_tassel.pl	1970-01-01 12:00:01.000000000 +1200
+++ quiet.patched/start_tassel.pl	2024-11-21 09:51:57.616549730 +1300
@@ -18,7 +18,7 @@
 }
 push(@fl, "$top/dist/sTASSEL.jar");
 my $CP = join(":", @fl);
-print $CP . "\n";
+# print $CP . "\n";
 
 # Scan @ARGV for Java memory arguments, and put rest in @args
 my $java_mem_min_default = "-Xms512m";
